components/logins/src/encryption.rs

```diff
@@ -75,14 +75,21 @@ impl EncryptorDecryptor {
     }
 }
 
-// public functions we expose over the FFI (which is why they take `String`
-// rather than the `&str` you'd otherwise expect)
-pub fn encrypt_string(key: String, cleartext: String) -> Result<String> {
-    EncryptorDecryptor::new(&key)?.encrypt(&cleartext)
+// Canary checking functions.  These are used to check if a key is still valid for a database.  The
+// basic process is:
+//   - When opening a database the first time, store the output of `create_canary()` alongside the DB
+//     and encryption key.
+//   - When reopening the database, check that the encryption key is still valid for the canary
+//     text using `check_canary()`
+//     - If it returns true, then it's safe to assume the key can decrypt the DB data
+//     - If it returns false, then the key is no longer valid.  It should be regenerated and the DB
+//       data should be wiped since we can no longer read it properly
+pub fn create_canary(text: &str, key: &str) -> Result<String> {
+    EncryptorDecryptor::new(key)?.encrypt(text)
 }
 
-pub fn decrypt_string(key: String, ciphertext: String) -> Result<String> {
-    EncryptorDecryptor::new(&key)?.decrypt(&ciphertext)
+pub fn check_canary(canary: &str, text: &str, key: &str) -> Result<bool> {
+    Ok(EncryptorDecryptor::new(&key)?.decrypt(&canary)? == text)
```

Comment:
I actually like these changes, they feel more natural as an API. However I wonder for the sake of consistency, we follow how autofill does so that it's exactly mirrored? We actually have the same `encrypt_string` `decrypt_string` in our encryption.rs 

Logins encryption.rs: https://github.com/mozilla/application-services/blob/logins-per-field-encryption/components/logins/src/encryption.rs#L78-L91

Autofill encryption.rs: https://github.com/mozilla/application-services/blob/c12b286312299d66a2404fa807d2a13f035095f6/components/autofill/src/encryption.rs#L74-L87

If we expose that similar to autofill's udl, it might end up being easier on the consumer kotlin/ios side?  

If not let me know, more of a thought than any kind of an opinion.

Reply:
That's a really good question.  From the Kotlin side, I think either would work and the coding effort is similar.  I don't have an opinion either way.

Long-term, I think that maybe we could do a quick refactor of both logins and autofill to make them more consistent.  I also wonder if at some point we should expose the `EncryptorDecryptor` directly rather than recreating it for each call.